com.unity.render-pipelines.high-definition/Runtime/Lighting/LightLoop/LightLoopDef.hlsl

```diff
@@ -169,164 +169,296 @@ float4 SampleEnv(LightLoopContext lightLoopContext, int index, float3 texCoord,
-    start = tileOffset;
-}
+    // This part (1) is loop-invariant. These values do not depend on the value of 'i'.
+    // They will only be computed once per category, not once per function call.
+    const uint tileBufferHeaderIndex = ComputeTileBufferHeaderIndex(tile, category, unity_StereoEyeIndex);
+    const uint tileRangeData         = TILE_BUFFER[tileBufferHeaderIndex]; // {last << 16 | first}
+    const bool isTileEmpty           = tileRangeData == UINT16_MAX;
 
-#ifdef USE_FPTL_LIGHTLIST
+    if (!isTileEmpty) // Avoid wasted work
+    {
+        const uint zBinBufferIndex0 = ComputeZBinBufferIndex(zBinRange[0], category, unity_StereoEyeIndex);
+        const uint zBinBufferIndex1 = ComputeZBinBufferIndex(zBinRange[1], category, unity_StereoEyeIndex);
+        const uint zBinRangeData0   = _zBinBuffer[zBinBufferIndex0]; // {last << 16 | first}
+        const uint zBinRangeData1   = _zBinBuffer[zBinBufferIndex1]; // {last << 16 | first}
+
+        // Recall that entities are sorted by the z-coordinate.
+        // So we can take the smallest index from the first bin and the largest index from the last bin.
+        // TODO: while it appears to work, for a large range, this code can actually fail. :-(
+        // Imagine that we are given the first and the last zBins, and that none of the lights overlap them.
+        // All the lights are in between. So the code below would give us an empty range, which is wrong.
+        // Should we loop over the entire zBinRange? Seems impractical... Maybe change the way we sort?
+        const uint2 tileEntityIndexRange = uint2(tileRangeData  & UINT16_MAX, tileRangeData  >> 16);
+        const uint2 zBinEntityIndexRange = uint2(zBinRangeData0 & UINT16_MAX, zBinRangeData1 >> 16);
+
+        if (IntervalsOverlap(tileEntityIndexRange, zBinEntityIndexRange)) // Avoid wasted work
+        {
+            const uint tileBufferBodyIndex = ComputeTileBufferBodyIndex(tile, category, unity_StereoEyeIndex);
 
-uint GetTileSize()
-{
-    return TILE_SIZE_FPTL;
-}
+            // The part (2) below will be actually executed during every function call.
+            while (i < n)
+            {
+                // Walk the list of entity indices of the tile.
+                uint tileEntityPair  = TILE_BUFFER[tileBufferBodyIndex + (i / 2)];        // 16-bit indices
+                uint tileEntityIndex = BitFieldExtract(tileEntityPair, 16 * (i & 1), 16); // First Lo, then Hi bits
+
+                // Entity indices are stored in the ascending order.
+                // We can distinguish 3 cases:
+                if (tileEntityIndex < zBinEntityIndexRange.x)
+                {
+                    i++; // Skip this entity; continue the (linear) search
+                    // Fine point: entities are currently sorted by Centroid.z.
+                    // So it could be that bin #0 starts with entity #1 (which is large)
+                    // while bin #1 contains both entity #0 (which is small) and entity #1.
```

Comment:
This is why I am wondering should we do a dedicated version for non range (lights) and another for range (volumetric ) ?

Reply:
I would prefer not to duplicate the code.